src/hotspot/share/jvmci/jvmciRuntime.cpp

```diff
@@ -827,62 +828,115 @@ void JVMCINMethodData::invalidate_nmethod_mirror(nmethod* nm) {
   }
 }
 
-JVMCIRuntime::JVMCIRuntime(int id) {
-  _init_state = uninitialized;
-  _shared_library_javavm = NULL;
-  _id = id;
-  _metadata_handles = new MetadataHandles();
-  JVMCI_event_1("created new JVMCI runtime %d (" PTR_FORMAT ")", id, p2i(this));
-}
-
 // Handles to objects in the Hotspot heap.
 static OopStorage* object_handles() {
   return Universe::vm_global();
 }
 
-jobject JVMCIRuntime::make_global(const Handle& obj) {
+jlong JVMCIRuntime::make_oop_handle(const Handle& obj) {
   assert(!Universe::heap()->is_gc_active(), "can't extend the root set during GC");
   assert(oopDesc::is_oop(obj()), "not an oop");
   oop* ptr = object_handles()->allocate();
-  jobject res = NULL;
-  if (ptr != NULL) {
-    assert(*ptr == NULL, "invariant");
+  jlong res = 0;
+  if (ptr != nullptr) {
+    assert(*ptr == nullptr, "invariant");
     NativeAccess<>::oop_store(ptr, obj());
-    res = reinterpret_cast<jobject>(ptr);
+    res = (jlong) ptr;
   } else {
     vm_exit_out_of_memory(sizeof(oop), OOM_MALLOC_ERROR,
                           "Cannot create JVMCI oop handle");
   }
-  MutexLocker ml(JVMCI_lock);
+  MutexLocker ml(_lock);
+  _oop_handles.append(ptr);
   return res;
 }
 
-void JVMCIRuntime::destroy_global(jobject handle) {
+bool JVMCIRuntime::probe_oop_handle(jlong handle, int index) {
+  oop* key = (oop*) handle;
+  if (key == _oop_handles.at(index)) {
+    _last_found_oop_handle_index = index;
+    return true;
+  }
+  return false;
+}
+
+int JVMCIRuntime::find_oop_handle(jlong handle) {
+  int len = _oop_handles.length();
+  int next = _last_found_oop_handle_index + 1;
+  int prev = MAX2(_last_found_oop_handle_index, 0) - 1;
+
+  // Search "outwards" from the index of the last found
+  // entry. Experimentation shows that this significantly
+  // reduces the amount of searching performed.
+  do {
+    if (next < len) {
+      if (probe_oop_handle(handle, next)) {
+        return next;
+      }
+      next++;
+    }
+    if (prev >= 0) {
+      if (probe_oop_handle(handle, prev)) {
+        return prev;
+      }
+      prev--;
+    }
+  } while (next - (prev + 1) < len);
+  return -1;
+}
+
+int JVMCIRuntime::release_and_clear_globals() {
+  int released = 0;
+  if (_oop_handles.length() != 0) {
+    // Collect non-null JNI handles into an array for
+    // the bulk release operation
+    ResourceMark rm;
+    oop** ptrs = NEW_RESOURCE_ARRAY(oop*, _oop_handles.length());
+    for (int i = 0; i < _oop_handles.length(); i++) {
```

Comment:
Now that _oop_handles is a vector of `oop*`, the resource allocation and
possibly some copying could be avoided. Do an in-place removal of nullptr and
clearing of the values pass over the vector, and then use
```
object_handles()->release(_oop_handles.at_ptr(0), _oop_handles.length())
```
For future, I wonder if it would be worth changing OopStorage bulk release to ignore nullptrs in the array, to better support this kind of use-case.  But I'm not sure how common this kind of use-case (array of handles, some of which might be nullptr) really might be.  I think there are only a couple uses of bulk release so far.

Reply:
Nice suggestion on avoiding resource allocation: https://github.com/openjdk/jdk/pull/8262/commits/385880c2bf0343aee8166c6bc45446e0033c876b

I don't think it's worth changing OopStorage bulk release for this use case.